java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java

```diff
@@ -0,0 +1,155 @@
+      Flowable<ChatCompletionChunk> flowable = service.streamChatCompletion(chatCompletionRequest);
+
+      AtomicBoolean isFirst = new AtomicBoolean(true);
+      AssistantMessage responseMessage = service.mapStreamToAccumulator(flowable)
+          .doOnNext(accumulator -> {
+            if (accumulator.isFunctionCall()) {
+              if (isFirst.getAndSet(false)) {
+                System.out.println("Executing function " + accumulator.getAccumulatedChatFunctionCall().getName() + "...");
+              }
+            } else {
+              if (isFirst.getAndSet(false)) {
+                System.out.print("Response: ");
+              }
+              if (accumulator.getMessageChunk().getContent() != null) {
+                System.out.print(accumulator.getMessageChunk().getContent());
+              }
+            }
+          })
+          .doOnComplete(System.out::println)
+          .lastElement()
+          .blockingGet()
+          .getAccumulatedMessage();
+      session.addMessage(responseMessage);
+
+      ChatFunctionCall functionCall = responseMessage.getFunctionCall();
+      if (functionCall != null) {
+        FunctionValidation<ChatMessage> functionValidation = session.validateFunctionCall(functionCall);
+        if (functionValidation.isValid()) {
+          log.info("Executing {} with arguments {}", functionCall.getName(), functionCall.getArguments().toPrettyString());
+          ChatMessage functionResponse = session.executeFunctionCall(functionCall, context);
+          log.info("Executed {} with results: {}", functionCall.getName(), functionResponse.getTextContent());
+          session.addMessage(functionResponse);
+        } else {
+          log.info("Function call {} failed.", functionCall.getName());
+          session.addMessage(new FunctionMessage("{\"error\": \"" + functionValidation.validationError().errorMessage() + "\"}", "error"));
+        }
+      } else {
+        System.out.print("Next Query: ");
+        String nextLine = scanner.nextLine();
+        if (nextLine.equalsIgnoreCase("exit")) {
+          System.exit(0);
+        }
+        ChatMessage nextMsg = new UserMessage(nextLine);
+        session.addMessage(nextMsg);
+      }
+    }
+  }
+
+  public static void main(String... args) throws Exception {
+    String systemPrompt = "You are an incredibly enthusiastic and joyous life coach. You provide advice to people. You look up any and all advice you give via the provided functions.";
+    Path toolsPath = Path.of("java", "apirag-starter", "src", "main", "resources", "tools", "advice.tools.json");
+    String openaiKey = System.getenv("OPENAI_API_KEY");
+    APIExecutor apiExecutor = new RESTExecutorFactory().create(new MapConfiguration(Map.of(
+        "type", "rest",
+        "url", "https://api.adviceslip.com"
+    )), "adviceapi");
+
+    LowLevelAgent agent = new LowLevelAgent(toolsPath, apiExecutor, systemPrompt, openaiKey);
+    Map<String, Object> context = Map.of("userid", 1);
```

Comment:
Remove the context from this example. It's not needed. Pass a Map.of() to the super class in the constructor of MyChatSession.

Reply:
ok, but this is one of the things that is hardest to understand. your implementation of ContextKeys in the AcornAgentServer is also not very easy to understand why the context is there for, since it's not often met in ChatBot examples